File: cpp-version/current/player.cpp
#pragma once

#include <vector>
#include <unordered_set>

#include "stat.h"
#include "level.h"

struct Player {
	Stat stat;
	const Level* level;
	std::vector<int> trace;
	Bitset neighbours;
	Bitset visited;
	Bitset previousVisited;

	Player& enter(const Level* level) {
		this->level = level;
		neighbours.set(level->entrance);
		return *this;
	}

	Player& visit(const int id) {
		return *this >> id;
	}

	Player& operator >> (const int id) {
		assert(neighbours.test(id));
		auto pr = level->vertex(id)->ToProbeStat(stat);
		assert(stat >= pr.req);
		stat += pr.diff;
		trace.push_back(id);
		neighbours += level->neighbours[id];
		neighbours -= level->excludeNeighbours[id];
		neighbours -= visited;
		previousVisited = visited;
		visited.set(id);
		return *this;
	}

	template<class T>
	Player& operator >> (const T t) {
		return *this >> level->id(t);
	}
};

std::ostream& operator << (std::ostream& os, const Player& p) {
	os << p.stat;
	os << "Neighbours: ";
	bool first = true;
	for (auto id : p.neighbours) {
		if (first) {
			first = false;
		} else {
			os << ", ";
		}
		os << p.level->vertex(id)->name;
	}
	os << "\n";
	os << "Trace: ";
	first = true;
	for (auto id : p.trace) {
		if (first) {
			first = false;
		} else {
			os << ", ";
		}
		os << p.level->vertex(id)->name;
	}
	os << "\n";
	return os;
}


void printTrace(const Stat& initStat, const Player& player) {
	Player p{initStat};
	p.enter(player.level);
	std::cout << p;
	std::cout << "--------------------------------------------------------------------------------\n";
	for (auto id : player.trace) {
		p >> id;
		std::cout << p;
		std::cout << "--------------------------------------------------------------------------------\n";
	}
	std::cout.flush();
}

File: src/drod/player.rs
use rust_dense_bitset::BitSet as _;
use rust_dense_bitset::DenseBitSet as BitSet;

use super::{Level, PlayerStat};

#[derive(Clone, Debug)]
pub(super) struct Player {
    pub(super) stat: PlayerStat,
    // level: Level,
    trace: Vec<u8>,
    neighbors: BitSet,
    visited: BitSet,
    pub(super) previous_visited: BitSet,
}

impl Player {
    pub(super) fn new() -> Self {
        Self {
            stat: PlayerStat::default(),
            // level: Level::new(),
            trace: Vec::new(),
            neighbors: BitSet::new(),
            visited: BitSet::new(),
            previous_visited: BitSet::new(),
        }
    }

    pub(super) fn with_stat(stat: PlayerStat) -> Self {
        Self {
            stat,
            ..Player::new()
        }
    }

    pub(super) fn enter(&mut self, level: &Level) {
        let idx = level.entrance as usize;
        self.neighbors.set_bit(idx, true)
    }

    pub(super) fn visit(&mut self, room_id: u8, level: &Level) {
        let idx = room_id as usize;
        assert!(self.neighbors.get_bit(idx));
        let probe = level.vertex(room_id).to_probe_stat(&self.stat.into());
        assert!(self.stat.ge(&probe.req));
        self.stat += probe.diff;
        self.trace.push(room_id);
        self.neighbors |= level.neighbors[idx];
        self.neighbors &= !level.excluded_neighbors[idx];
        self.neighbors &= !self.visited;
        self.previous_visited = self.visited;
        self.visited.set_bit(idx, true);
    }
}
